src/gst.rs

```diff
@@ -84,21 +98,47 @@ impl RtspServer {
         Ok(maybe_app_src)
     }
 
-    pub fn set_credentials(&mut self, user_pass: Option<(&str, &str)>) -> Result<()> {
-        let auth = user_pass.map(|(user, pass)| {
-            let auth = RTSPAuth::new();
-            /*
-            let perm = RTSPToken::new(
-                ...
-            );
-            auth.add_basic(RTSPAuth::make_basic(user, pass).as_str(), &perm);
-            */
-            // TODO TLS https://thiblahute.github.io/GStreamer-doc/gst-rtsp-server-1.0/rtsp-server.html?gi-language=c
-            auth
-        });
+    pub fn set_credentials(&self, user :&str, pass :&str) -> Result<()> {
+        let auth = match self.server.get_auth() {
+            Some(x) => x,
+            None =>  RTSPAuth::new(),
+        };
+
+        if ! user.is_empty() && ! pass.is_empty() {
+            info!("Setting credentials for user {}", user);
+            debug!("Password is {}", pass);
+            let token = RTSPToken::new(&[(*RTSP_TOKEN_MEDIA_FACTORY_ROLE, &"watcher")]);
+            let basic = RTSPAuth::make_basic(user, pass);
+            auth.set_supported_methods(RTSPAuthMethod::Basic);
+            auth.add_basic(basic.as_str(), &token);
+
+            // Now we have basic auth we stop others from watching
+            let mut token = RTSPToken::new_empty();
+            auth.set_default_token(Some(&mut token));
+        } else {
+            let mut token = RTSPToken::new(&[(*RTSP_TOKEN_MEDIA_FACTORY_ROLE, &"watcher")]);
+            auth.set_default_token(Some(&mut token)); //By default anyone can watch if we don't turn on basic
+        }
 
-        self.server.set_auth(auth.as_ref());
+        self.server.set_auth(Some(&auth));
+        Ok(())
+    }
+
+    pub fn set_tls(&self, cert_file: &str, client_auth: TlsAuthenticationMode) -> Result<()> {
+        if ! cert_file.is_empty() {
+            info!("Setting up TLS using {}", cert_file);
+            let auth = match self.server.get_auth() {
+                Some(x) => x,
+                None =>  RTSPAuth::new(),
+            };
+
+            let cert = TlsCertificate::new_from_pem(&fs::read_to_string(cert_file).expect("TLS file not found")).expect("Not a valid TLS certificate");
```

Comment:
Looks like there's `new_from_file` which will avoid having to slurp the whole thing into a string.

Reply:
Ok, but this means that the user won't get a different message if its a missing file or if incorrect format.

Reply:
Oh, that's a good point, especially relevant with TLS certs. The way you have it will be fine; can you add a comment explaining this?

Reply:
Ok changed back being extra verbose about the reason